pkg/apis/meta.go

```diff
@@ -0,0 +1,303 @@
+const AnnoKeyOriginPath = "grafana.app/originPath"
+const AnnoKeyOriginKey = "grafana.app/originKey"
+const AnnoKeyOriginTimestamp = "grafana.app/originTimestamp"
+
+// ResourceOriginInfo is saved in annotations.  This is used to identify where the resource came from
+// This object can model the same data as our existing provisioning table or a more general git sync
+type ResourceOriginInfo struct {
+	// Name of the origin/provisioning source
+	Name string `json:"name,omitempty"`
+
+	// The path within the named origin above (external_id in the existing dashboard provisioing)
+	Path string `json:"path,omitempty"`
+
+	// Verification/identification key (check_sum in existing dashboard provisioning)
+	Key string `json:"key,omitempty"`
+
+	// Origin modification timestamp when the resource was saved
+	// This will be before the resource updated time
+	Timestamp *time.Time `json:"time,omitempty"`
+
+	// Avoid extending
+	_ any `json:"-"`
+}
+
+// Accessor functions for k8s objects
+type GrafanaResourceMetaAccessor interface {
+	GetUpdatedTimestamp() (*time.Time, error)
+	SetUpdatedTimestamp(v *time.Time)
+	GetCreatedBy() string
+	SetCreatedBy(user string)
+	GetUpdatedBy() string
+	SetUpdatedBy(user string)
+	GetFolder() string
+	SetFolder(uid string)
+	GetSlug() string
+	SetSlug(v string)
+	GetTitle() string
+	SetTitle(v string)
+	GetOriginInfo() (*ResourceOriginInfo, error)
+	SetOriginInfo(info *ResourceOriginInfo)
+	GetOriginName() string
+	GetOriginPath() string
+	GetOriginKey() string
+	GetOriginTimestamp() (*time.Time, error)
+}
+
+var _ GrafanaResourceMetaAccessor = (*grafanaResourceMetaAccessor)(nil)
+
+type grafanaResourceMetaAccessor struct {
+	obj runtime.Object
+}
+
+func MetaAccessor(obj runtime.Object) GrafanaResourceMetaAccessor {
+	return &grafanaResourceMetaAccessor{obj}
+}
+
+func (m *grafanaResourceMetaAccessor) set(key string, val string) {
+	acc, err := meta.Accessor(m.obj)
+	if err != nil {
+		return
+	}
+
+	anno := acc.GetAnnotations()
+	if val == "" {
+		if anno != nil {
+			delete(anno, key)
+		}
+	} else {
+		if anno == nil {
+			anno = make(map[string]string)
+		}
+		anno[key] = val
+	}
+	acc.SetAnnotations(anno)
+}
+
+func (m *grafanaResourceMetaAccessor) get(key string) string {
+	acc, err := meta.Accessor(m.obj)
+	if err != nil {
+		return ""
+	}
+
+	return acc.GetAnnotations()[key]
+}
+
+func (m *grafanaResourceMetaAccessor) GetUpdatedTimestamp() (*time.Time, error) {
+	acc, err := meta.Accessor(m.obj)
+	if err != nil {
+		return nil, err
+	}
+
+	v, ok := acc.GetAnnotations()[AnnoKeyUpdatedTimestamp]
+	if !ok || v == "" {
+		return nil, nil
+	}
+	t, err := time.Parse(time.RFC3339, v)
+	if err != nil {
+		return nil, fmt.Errorf("invalid updated timestamp: %s", err.Error())
+	}
+	return &t, nil
+}
+
+func (m *grafanaResourceMetaAccessor) SetUpdatedTimestampMillis(v int64) {
+	if v > 0 {
+		t := time.UnixMilli(v)
+		m.SetUpdatedTimestamp(&t)
+	} else {
+		m.SetUpdatedTimestamp(nil)
```

Comment:
Is this correct? In what situation would we set the UpdatedTimestamp to nil, as opposed to ignoring the input if it's 0? (sorry if this just is a k8s thing I don't know)

Reply:
updated to make it a bit more clear -- the updatedTimestamp is a value in a map[string]string -- so we need to trigger removing it from the right key (and that all exists in the set nil case)